ffmpeg/lpms_ffmpeg.c

```diff
@@ -753,6 +797,37 @@ static int open_video_decoder(input_params *params, struct input_ctx *ctx)
       AVCodec *c = avcodec_find_decoder_by_name("h264_cuvid");
       if (c) codec = c;
       else fprintf(stderr, "Cuvid decoder not found; defaulting to software\n");
+    } else if (AV_CODEC_ID_H264 == codec->id &&
+                       AV_HWDEVICE_TYPE_VAAPI == params->hw_type) {
+      enum AVHWDeviceType type;
+      type = av_hwdevice_find_type_by_name("vaapi");
+      if (type == AV_HWDEVICE_TYPE_NONE) {
+        fprintf(stderr, "Device type vaapi is not supported.\n");
+        fprintf(stderr, "Available device types:");
+        while((type = av_hwdevice_iterate_types(type)) != AV_HWDEVICE_TYPE_NONE)
+          fprintf(stderr, " %s", av_hwdevice_get_type_name(type));
+        fprintf(stderr, "\n");
+        ret = -1;
+        goto open_decoder_err;
+      }
+
+      AVCodec *c = NULL;
+      int i;
+      ret = av_find_best_stream(ic, AVMEDIA_TYPE_VIDEO, -1, -1, &c, 0);
+      if (c) {
+        codec = c;
+        for (i = 0;; i++) {
+          const AVCodecHWConfig *config = avcodec_get_hw_config(codec, i);
+          if (!config) {
+            dd_err("Decoder %s does not support device type %s.\n", codec->name, av_hwdevice_get_type_name(type));
+          }
+          if (config->methods & AV_CODEC_HW_CONFIG_METHOD_HW_DEVICE_CTX &&
+                  config->device_type == type) {
+            ctx->hw_pix_fmt = config->pix_fmt;
+            break;
+          }
+        }
+      } else fprintf(stderr, "Vaapi decoder not found; defaulting to software\n");
```

Comment:
Is this log on the correct line? I don't think the `c` (or `codec`) as returned by `find_best_stream` has any relationship to the vaapi decoder? Can you elaborate on the expected behavior here?

As a matter of style, generally if there's a long block of code followed by brief error handling in a separate branch, then I like to put the error handling first, which makes it easier to visually scan. For example:

```c
if (!c) fprintf(stderr, "Vaapi decoder not found ... ");
else {
  codec = c;
  // ... long block of code ...
}
```

Reply:
Reverted the change.

For error handling would keep a note.